squire_lib/src/pairings/greedy.rs

```diff
@@ -15,45 +13,36 @@ pub fn greedy_pairings<Players>(
 where
     Players: IntoIterator<Item = PlayerId>,
 {
-    let plyrs_iter = plyrs.into_iter();
-    let size = plyrs_iter.size_hint().pipe(|(a, b)| b.unwrap_or(a));
-    let mut plyrs = VecDeque::with_capacity(size);
-    for id in plyrs_iter {
-        plyrs.push_front(id);
-    }
+    let mut plyrs: VecDeque<_> = plyrs.into_iter().collect();
     let mut digest = Pairings {
         paired: Vec::with_capacity(plyrs.len() / match_size + 1),
         rejected: Vec::new(),
     };
     while plyrs.len() >= match_size {
-        let Some(first) = plyrs.pop_back() else { break; };
-        let mut index_buffer: Vec<usize> = Vec::with_capacity(match_size);
+        let Some(first) = plyrs.pop_front() else { break; };
         let mut id_buffer: Vec<PlayerId> = Vec::with_capacity(match_size);
 
-        for (i, plyr) in plyrs.iter().enumerate().rev() {
+        for plyr in &plyrs {
             let current_pairing = std::iter::once(&first).chain(id_buffer.iter());
             if valid_pairing(opps, current_pairing, plyr, repair_tol) {
-                index_buffer.push(i);
                 id_buffer.push(*plyr);
-                if index_buffer.len() == match_size {
+                if id_buffer.len() == match_size {
                     break;
                 }
             }
         }
 
-        if index_buffer.len() == match_size {
+        if id_buffer.len() == match_size {
+            plyrs.retain(|p| !id_buffer.contains(p));
             let mut pairing: Vec<PlayerId> = Vec::with_capacity(match_size);
             pairing.push(first);
-            for i in index_buffer.into_iter() {
-                let id = plyrs.remove(i).unwrap();
-                pairing.push(id);
-            }
+            pairing.extend(id_buffer);
```

Comment:
Do we need `pairing` at all now? Before, `pairing` was populated with the elements pulled out from `index_buffer`, but `retain` removes them now. I think we can simply insert `first` into `id_buffer` and push `id_buffer` into `digest`.